src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,350 @@
+/-
+Copyright (c) 2020 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Adam Topaz.
+-/
+
+import ring_theory.algebra
+import linear_algebra
+
+/-!
+# Tensor Algebras
+
+Given a commutative ring R, and an R-module M, we construct the tensor algebra of M.
+This is the free R-algebra generated by the module M.
+
+## Notation
+
+1. `tensor_algebra R M` is the tensor algebra itself.
+  It is endowed with an R-algebra structure.
+2. `univ R M` is the R-linear map `M → tensor_algebra R M`.
+3. Given a linear map `f : M → A` to an R-algebra A,
+  `lift R M f` is the lift of `f` to an R-algebra morphism
+  `tensor_algebra R M → A`.
+
+## Theorems
+
+1. `univ_comp_lift` states that the composition
+  `(lift R M f) ∘ (univ R M)` is identical to `f`.
+2. `lift_unique` states that whenever an R-algebra morphism
+  `g : tensor_algebra R M → A` is given whose composision with `univ R M` is `f`,
+  then one has `g = lift R M f`.
+-/
+
+universes u1 u2 u3
+
+variables (R : Type u1) [comm_ring R]
```

Comment:
We have tensor products of semimodules... How much of this file works in that generality? I can imagine that secretly sometimes we might be interested in that case, as mathematicians. Even though we never really talk or write about it.
(Note: I didn't think about this at all. I suggest you just ask lean. If lots of stuff breaks in a nontrivial way... then keep it like it was.)

Reply:
Aren't algebras only defined for commutative rings?

Reply:
Oh, they're defined for comm_semirings, so all of this should just work in that level of generality.